<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">

<!-- Created on Jul 1, 2005 4:34:02 PM -->

<html>

   <head>
      <title>
         IntAct Q&amp;A User #05
      </title>
      <meta name="GENERATOR" content="Arachnophilia 5.2">
      <meta name="FORMATTER" content="Arachnophilia 5.2">
      <link rel="stylesheet" type="text/css" href="intactdjl.css">
   </head>

   <body>
      <table border="0" height="100%" width="100%">
         <tr>
            <td valign="top" rowspan="2" height="100%" width="1%">
               <table class="left" height="100%" width="1%">
                  <tr>
                    <td align="top" valign="top">
                        <a href="http://www.ebi.ac.uk/intact/index.jsp" target="_blank" title="Opens new window"><img border="0" src="images/logo_intact.png"></a>
                        <br>
                        <br>
                        <p class="links">
                        &ordm;&nbsp;
			<a class="links" href="http://www.ebi.ac.uk/intact/index.jsp" target="_blank" title="Opens new window">IntAct</a>
                        <br>
                        <nobr>&ordm;&nbsp;
			<a class="links" href="http://www.ebi.ac.uk/intact/search/do/initAdvSearch" target="_blank" title="Opens new window">Advanced Search</a></nobr>
                        </p>
                     </td>
                  </tr>
               </table>
            </td>
            <td valign="top" height=100%>
               <table border="0" height="100%" width="100%">
                  <tr>
                     <td class="header" height="30" width="100%" colspan="2">
                        <a class="header" href="index.html">&nbsp;&nbsp;Help &amp; Tutorial Website</a>
                     </td>
                  </tr>
                  <tr>
                     <td height="5" colspan="2" valign="top">
                        <hr>
                     </td>
                  </tr>
                  <tr>
                     <td width="50%" height="18" valign="top" align="LEFT">
                        <a href="index.html" title="IntAct Q&A Homepage">Home</a>
                        &nbsp;&nbsp;&#47;&nbsp;&nbsp;
                        <a href="IntActQ&A_User_Contents.html">User_Contents</a>
                     </td>
                  </tr>
                  <tr>
                     <td height="5" colspan="2" align="left" valign="top">
                        <center>
                           <hr>
                        </center>
                     </td>
                  </tr>
                  <tr>
                     <td colspan="2" align="left" valign="center">
                     </td>
                  </tr>
                  <tr>
                     <td colspan="2" align="center" valign="top" height="100%">
                        <table class="inset" width="75%" height="75%">
                           <tr>
                              <td height="33" class="inset">
                                 <div align="CENTER">
                                    <b>What do the results mean?</b>
                                    <br>
                                 </div>
                              </td>
                           </tr>
                           <tr>
                              <td class="inset" height="368" colspan="2" valign="TOP">
                                 <p>
                                 The image below shows the IntAct homepage, the starting point for most users. As an example, the protein lsm7 will be investigated.
                                 </p>
                                 <p class="centeredImage">
                                 <img border="0" class="screengrab" src="images/User_05_pic01.png" width="690" height="522">
                                 </p>
                                 <p>
                                 The output above shows that there are three entries in the IntAct database for proteins that have lsm7 in their name (second column from 
                                 the left) and would appear to be 
                                 <a href="http://www.mblogic.net/glossary/default.asp?txtSearch=ortholog&field=Term&Button=Find&lang=" title="ortholog definition" target="_blank">orthologues</a>. On the right, next to the protein name is the IntAct Accession number. To get the next page I clicked on &quot;EBI-141&quot;.
                                 </p>
                                 <p class="centeredImage">
                                 <img border="0" class="screengrab"  src="images/User_05_pic02.png" width="690" height="522">
                                 </p>
                                 <p class="centeredImage">
                                 <img border="0" class="screengrab"  src="images/User_05_pic03.png" width="690" height="522">
                                 <br>
                                 <br>
                                 <p>
                                 The page above shows a variety of information about the protein lsm7_yeast. On the right of the name is the IntAct Accession number. 
                                 Below that is the source organism, a brief description of the putative protein function and any alternative gene names (in this case, 
                                 there are none). Below that is a list of alternative gene names for the protein and count of the number of interactions the protein 
                                 participates, at the current time then are a number of cross-references (Xref) to entries in other databases. First there is a link to the entry 
                                 for lsm7 in the Saccharomyces Genome Database (SGD) followed by a number of Gene Ontology IDs which describe the protein (GO). Remember 
                                 Gene Ontology is hierarchical and some Gene Ontology IDs represent very general descriptions of function. For more information about 
                                 Gene Ontology click
                                 <a href="http://www.geneontology.org/GO.doc.shtml" title="An Introduction to the Gene Ontology" target="_blank">here</a>. The other two 
                                 links are for 
                                 <a href="http://www.ebi.uniprot.org/index.shtml" title="Welcome to UniProt" target="_blank">UniProtKB</a>
                                 and 
                                 <a href="http://www.ebi.ac.uk/interpro/" title="InterPro Home" target="_blank">InterPro</a>. UniProtKB (Universal Protein 
                                 Resource - Knowledgebase) is a protein databank for proteins sequence and function and InterPro displays information regarding protein families, 
                                 domains and functional sites. Also featured is the protein sequence and the number of amino acids it. 
                                 </p>
                                 <p>
                                 To see what proteins lsm7_yeast interacts with I clicked on "see all interaction partners" to get the next page.
                                 </p>
                                 <p class="centeredImage">
                                 <img border="0" class="screengrab"  src="images/User_05_pic04.png" width="690" height="522">
                                 </p>
                                 <p>
                                 The page above shows a list of the proteins that interact with lsm7_yeast. Note that the 
                                 <u>Number of interactions</u>
                                 refers to the number of experiments which feature the given protein and not the number of proteins with which the given protein has been 
                                 shown to interact with. It is possible to get a graphical representation of the interaction network with lsm7 by ticking the box next to 
                                 <b><u><font color="#FF0000">lsm7_yeast</font></u></b>, this will be shown later. To see the experiments which feature lsm7 I clicked the 
                                 <u>14</u>
                                 to get the following page.
                                 </p>
                                 <p class="centeredImage">
                                 <img border="0" class="screengrab"  src="images/User_05_pic05.png" width="690" height="522">
                                 </p>
                                 <p>
                                 The page above shows the seven experiments that feature lsm7_yeast. A brief summary is shown including a description and the 
                                 total number of proteins reported from the experiment. To look at one Experiment I clicked on ga-205 to get the next page 
                                 (clicking on EBI-139 would also go to the same page)
                                 </p>
                                 <p class="centeredImage">
                                 <img border="0" class="screengrab"  src="images/User_05_pic06.png" width="690" height="522">
                                 </p>
                                 <p one of main aims in designed the intact database is to provide as much information about protein-protein interactions possible. >
                                 </p>
                                 <p>
                                 In the Experiment sections,  the details  are given including the experimental procedure used to detect the interaction (Interaction 
                                 identification) and identify the protein involved (Participant identification). Both of these have links that give a description of the 
                                 method, the PSI-MI ID and a link to a PubMed reference which consists of a relevant article that explains the technique (see below). 
                                 The Xref links refer to the published article from which the data has been produced. 
                                 </p>
                                 <p>
                                 The Interaction section gives the IntAct name for the experiment and the IntAct ID (in this case ga-205 and EBI-141 respectively). The 
                                 type of interaction is given with a link to another page that gives a definition and the corresponding PSI-MI ID.
                                 </p>
                                 <p class="centeredImage">
                                 <img border="0" class="screengrab"  src="images/User_05_pic07.png" width="690" height="522">
                                 </p>
                                 <p>
                                 To access the page below, I clicked on the Xref pubmed link 10504710. The page below shows the PubMed entry for the article that explains 
                                 experimental technique, &#8220;tandem affinity purification&#8221;, used to investigate proteins interactions.
                                 </p>
                                 <p class="centeredImage">
                                 <img border="0" class="screengrab"  src="images/User_05_pic08.png" width="690" height="522">
                                 </p>
                                 <p class="centeredImage">
                                 <img border="0" class="screengrab"  src="images/User_05_pic06.png" width="690" height="522">
                                 </p>
                                 <p>
                                 Returning back to the experiment page for ga-205 (or EBI-139), note that each experiment reports on multiple interactions. Thus many 
                                 interactions belong to the experiment gavin-2002 (EBI-12) but each interaction (in this case EBI-139) refers to only one protein (in 
                                 this case lsm7_yeast) as bait for investigating interaction partners.
                                 </p>
                                 <p>
                                 To view the interacting partners of lsm7_yeast using a network visualization tool tick the box next to &quot;Interaction&quot; and click 
                                 on &quot;Graph&quot; to get the page below.
                                 </p>
                                 <p class="centeredImage">
                                 <img border="0" class="screengrab"  src="images/User_05_pic09.png" width="690" height="522">
                                 </p>
                                 <p>
                                 HierachView shows an interaction network using nodes to represent individual proteins and inter-connecting lines to show interactions. 
                                 The network can be expanded by selecting &quot;center the view&quot; in the left hand side frame and clicking on a node (protein). I 
                                 clicked on the node lsm4_yeast (~10 o'clock on the network above) to expand the network and give the page below.
                                 </p>
                                 <p class="centeredImage">
                                 <img border="0" class="screengrab"  src="images/User_05_pic10.png" width="690" height="522">
                                 </p>
                                 <p>
                                 The result in an expansion of the network into a more complicated network by adding more interactions stored in the database. On selecting 
                                 a protein (lsm4_yeast) the corresponding Gene Ontology IDs are given in the right hand side column. These Gene Ontology IDs describe the 
                                 protein properties. For more information about Gene Ontology click 
                                 <a href="IntActQ&A_User_03.html" title="What is gene ontology (GO) and why use it">here</a>
                                 or visit the gene ontology database 
                                 <a href="http://www.godatabase.org" title="The Gene Ontology Database - AmiGO!" target="_blank">http://www.godatabase.org</a>. Interaction 
                                 network data can be downloaded and viewed in a third party network visualisation program. There are many such programs available, for a 
                                 list and further information click 
                                 <a href="IntActQ&A_User_10.html#NetVizSoftware" title="Analyzing IntAct Data in Third Party Network Visualization Software">here</a>. 
                                 </p>
                                 <p>
                                 To download the data that produces the HierachView graph simply click on the &quot;Download&quot; button under &quot;PSI-XML&quot; to get 
                                 a page similar to that below.
                                 </p>
                                 <p class="centeredImage">
                                 <img border="0" class="screengrab"  src="images/User_05_pic11.png" width="690" height="522">
                                 </p>
                                 <p>
                                 The page above shows the data from IntAct in
                                 <a href="http://en.wikipedia.org/wiki/XML" title="XML info @ Wikipedia" target="_blank">XML</a>
                                 format called 
                                 <a href="http://psidev.sourceforge.net/mi/xml/doc/user/" title="PSI-MI HUPO" target="_blank">PSI-MI.</a>
                                 PSI-MI which was developed in collaboration with the Human Proteome Organisation (
                                 <a href="http://www.hupo.org/" target="_blank">HUPO</a>
                                 ) 
                                 to enable the exchange of molecular interaction data using XML.
                                 </p>
                                 <p>
                                 <b>To save the IntAct data</b>
                                 </p>
                                 <p>
                                 * Using 
                                 <a href="http://www.mozilla.org/products/firefox/" title="Mozzila's Firefox FREE web-broswer - rediscover the web" target="_blank">Firefox</a>
                                 <br>
                                 File / Save Page As / - ensure &quot;Save type as &quot; is selected as &quot;XML Document&quot; / Save
                                 <br>
                                 <br>
                                 * Using IE
                                 <br>
                                 File / Save  As... / - ensure &quot;Save  as type:&quot; is selected as &quot;XML file (*.xml)&quot; / Save
                                 </p>
                              </td>
                           </tr>
                        </table>
                     </td>
                  </tr>
                  <tr>
                     <td colspan="2" valign="bottom" height=5%>
                        <table align="bottom" border="0" width="100%" height="20">
                           <tr>
                              <td height="25" colspan="2" align="left">
                                 <hr>
                              </td>
                           </tr>
                           <tr>
                              <div class="footer">
                                 <td class="footer">
                                    <p class="footer">
                                    Please send any questions or suggestions to 
                                    <a class="footer" href="mailto:intact-help@ebi.ac.uk">intact-help@ebi.ac.uk</a>
                                    <br>
                                    Last modified:
                                    <i>11-May-2005 10:28 PM by dlegge</i>
                                    </p>
                                 </td>
                              </div>
                           </tr>
                        </table>
                     </td>
                  </tr>
               </table>
            </td>
         </tr>
      </table>
   </body>
</html>
